sm/pmp.c

```diff
@@ -223,14 +223,23 @@ static void send_and_sync_pmp_ipi(int region_idx, enum ipi_type type, uint8_t pe
     }
   }
 
+  // wait until all events have been handled.
+  // prevent deadlock by consuming incoming IPIs.
+  uint32_t incoming_ipi = 0;
   /* wait until every other hart sets PMP */
   for(uintptr_t i=0, m=mask; m; i++, m>>=1) {
     if(m & 1) {
       while( atomic_read(&ipi_mailbox[i].pending) ) {
-        continue;
+        incoming_ipi |= atomic_swap(HLS()->ipi, 0);
       }
     }
   }
+  
+  // if we got an IPI, restore it; it will be taken after returning
+  if (incoming_ipi) {
+    *HLS()->ipi = incoming_ipi;
+    mb();
+  }
```

Comment:
So this feels slightly sketchy.
We're marking all incoming IPIs as "handled" as far as the other harts are concerned, but then re-marking them as pending and fully returning.
It SHOULD result in us immediately trapping to M-mode and handling the IPI.
But, it seriously complicates our security model. When a hart sees a PMP IPI has been handled by other cores, we were previously able to make a strong assumption about the security state globally. That is no longer true, though probably safe.
Can we instead handle these IPIs _before_ we exit M-mode? That would then hold the same guarantee as before I believe.

Reply:
Whoops, ya it was supposed to handle the ipi calls in that if statement. I was going to work on the ipi handler so that it is an actual c function soon. This was just a temporary fix to get multithreading working.